File: PrimaMateria/PrimaMateria/Classes/XTRColorPickerViewController.m
//
//  XTRColorPickerViewController.m
//  PrimaMateria
//
//  Created by Jerry Porter on 5/11/14.
//  Copyright (c) 2016 xTrensa. All rights reserved.
//

@interface XTRColorPickerViewController ()

@end

@implementation XTRColorPickerViewController

- (IBAction)selectColor: (id)sender {
    NSArray *values = [NSArray arrayWithObjects: @(self.redSlider.value), @(self.greenSlider.value), @(self.blueSlider.value), self.colorTitle.text, nil];
    NSArray *keys = [NSArray arrayWithObjects: RED_COLOR_COMPONENT, GREEN_COLOR_COMPONENT, BLUE_COLOR_COMPONENT, SERIES_COLOR_KEY, nil];
    NSDictionary *dict = [NSDictionary dictionaryWithObjects: values forKeys: keys];
    [self dismissViewControllerAnimated: YES completion: nil];
	[[NSNotificationCenter defaultCenter] postNotificationName: NOTIFICATION_COLOR_SELECTED object: dict];
}

- (void)presetSlidersWithColor: (UIColor *)aColor {
    [self.redSlider setValue: aColor.red /*clr[0]*/ animated: YES];
    [self.greenSlider setValue: aColor.green /*clr[1]*/ animated: YES];
    [self.blueSlider setValue: aColor.blue /*clr[2]*/ animated: YES];
    self.previewView.backgroundColor = [UIColor colorWithRed: self.redSlider.value green: self.greenSlider.value blue: self.blueSlider.value alpha: 1.0];
    [self performSelector: @selector(redSliderValueChanged:) withObject: self.redSlider afterDelay: 0.0];
}

- (void)redSliderValueChanged: (UISlider *)slider {
    [UIView beginAnimations: nil context: NULL];
    self.previewView.backgroundColor = [UIColor colorWithRed: slider.value green: self.greenSlider.value blue: self.blueSlider.value alpha: 1.0];
    [UIView commitAnimations];
}

- (void)greenSliderValueChanged: (UISlider *)slider {
    [UIView beginAnimations: nil context: NULL];
    self.previewView.backgroundColor = [UIColor colorWithRed: self.redSlider.value green: slider.value blue: self.blueSlider.value alpha: 1.0];
    [UIView commitAnimations];
}

- (void)blueSliderValueChanged: (UISlider *)slider {
    [UIView beginAnimations: nil context: NULL];
    self.previewView.backgroundColor = [UIColor colorWithRed: self.redSlider.value green: self.greenSlider.value blue: slider.value alpha: 1.0];
    [UIView commitAnimations];
}

- (void)dealloc {
    self.redSlider = nil;
    self.greenSlider = nil;
    self.blueSlider = nil;
    self.previewView = nil;
    self.colorTitle = nil;
}

- (BOOL) shouldAutorotateToInterfaceOrientation: (UIInterfaceOrientation) interfaceOrientation {
    return UIInterfaceOrientationIsLandscape(interfaceOrientation);
}

@end

File: PrimaMateria/PrimaMateria/Classes/XTRColorPickerViewController.swift
//
//  XTRColorPickerViewController.swift
//  PrimaMateria
//
//  Created by Jerry Porter on 4/14/16.
//  Copyright ©2018 xTrensa. All rights reserved.
//

class XTRColorPickerViewController : UIViewController {
    
    @IBOutlet var redSlider : UISlider!
    @IBOutlet var greenSlider : UISlider!
    @IBOutlet var blueSlider : UISlider!
    @IBOutlet var alphaSlider : UISlider!
    @IBOutlet var colorTitle : UILabel!
    @IBOutlet var previewView: UIView!
    
    var seriesName : String?
    
    // MARK: - Initialization Methods
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)!
    }
    
    // MARK: - Internal Methods
    
    // MARK: - Action Methods
    
    @IBAction func selectColor(_ sender: UIButton) {
        let values = [redSlider.value, greenSlider.value, blueSlider.value, alphaSlider.value, colorTitle.text!] as [Any]
        let keys = [ColorComponent.red, ColorComponent.green, ColorComponent.blue, ColorComponent.alpha, SERIES_COLOR_KEY]
        let dict = zip(keys, values).reduce([String : Any]()){ var d = $0; d[$1.0] = $1.1; return d }
        
        dismiss(animated: true, completion:  nil)
        NotificationCenter.default.post(name: .colorSelectedNotification, object: dict)
    }
    
    func presetSlidersWithColor(_ aColor: UIColor) {
        redSlider.setValue(Float(aColor.red()), animated:  true)
        greenSlider.setValue(Float(aColor.green()), animated: true)
        blueSlider.setValue(Float(aColor.blue()), animated: true)
        alphaSlider.setValue(Float(aColor.alpha()), animated: true)
        
        previewView.backgroundColor = UIColor(red: CGFloat(redSlider.value), green: CGFloat(greenSlider.value), blue: CGFloat(blueSlider.value), alpha: CGFloat(alphaSlider.value))
    }
    
    @IBAction func redSliderValueChanged(_ sender: UISlider) {
        UIView.beginAnimations(nil, context: nil)
        previewView.backgroundColor = UIColor(red: CGFloat(sender.value), green: CGFloat(greenSlider.value), blue: CGFloat(blueSlider.value), alpha: CGFloat(alphaSlider.value))
    }
    
    @IBAction func greenSliderValueChanged(_ sender: UISlider) {
        UIView.beginAnimations(nil, context: nil)
        previewView.backgroundColor = UIColor(red: CGFloat(redSlider.value), green: CGFloat(sender.value), blue: CGFloat(blueSlider.value), alpha: CGFloat(alphaSlider.value))
    }
    
    @IBAction func blueSliderValueChanged(_ sender: UISlider) {
        UIView.beginAnimations(nil, context: nil)
        previewView.backgroundColor = UIColor(red: CGFloat(redSlider.value), green: CGFloat(greenSlider.value), blue: CGFloat(sender.value), alpha: CGFloat(alphaSlider.value))
    }
    
    @IBAction func alphaSliderValueChanged(_ sender: UISlider) {
        UIView.beginAnimations(nil, context: nil)
        previewView.backgroundColor = UIColor(red: CGFloat(redSlider.value), green: CGFloat(greenSlider.value), blue: CGFloat(blueSlider.value), alpha: CGFloat(sender.value))
    }
    
    // MARK: - View Management Methods
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let aColor = XTRColorFactory.colorForString(seriesName!)
        colorTitle.text = seriesName!
        previewView.backgroundColor = aColor
        presetSlidersWithColor(aColor)
    }
    
    override var shouldAutorotate : Bool {
        return false
    }
    
    override var supportedInterfaceOrientations : UIInterfaceOrientationMask {
        return .landscape
    }
    
    // MARK: - Memory Management Methods
    
    deinit {
        redSlider = nil
        greenSlider = nil
        blueSlider = nil
        alphaSlider = nil
        previewView = nil
        colorTitle = nil
    }
    
}
